### PR TITLE
Add Swarm multicodecs

### PR DESCRIPTION
**Pull request by justelad**
Hi,

I'd like to add these two Swarm primitive types that we'd like to be able to reference from ENS resolver contracts.

Please let me know if I should shift the codes around (although obviously lower numbers = less data on chain).

Cheers

### CODE REVIEW

**Comment by Stebalien** [comment]
So we can correctly categorize these, how are they being used?

**Comment by justelad** [comment]
@Stebalien These are Swarm notions of base layer primitives that resolve automagically to actual content.

My assumption is that users would want to point to either one of them when setting an ENS resolver contract content hash as discussed in EIP-1577.

So if I understand the question, they will be used by users to point to content. 
In general I don't see any need to represent any other Swarm notions inside multicodecs as we define file metadata inside manifests; not as a self defining hash. So in fact the purpose of this PR is to define the standard of how to set content hashes in the revised ENS resolvers.

**Comment by justelad** [comment]
Also, if I may ask to put us in an address space where we could have a few more subsequent future entries reserved.

Many thanks

**Comment by justelad** [comment]
Should I revise anything in this PR?

**Comment by Stebalien** [comment]
So, are these used to identify the type of content? I'm asking because I _think_ these should be categorized as IPLD codecs: they tell you how to understand the addressed content (which I'm guessing isn't self-describing).

**Comment by Stebalien** [comment]
On the original swarm PR, there was a lot of discussion around using CIDs. That is, ENS records would have values of the format `<namespace (multicodec)><CID><path?>`. For example, `<swarm-ns><CID (codec=swarm-manifest)>/path/to/file`.

**Comment by justelad** [comment]
@Stebalien I suspect I'm not 100% following you and I'm not 100% sure how CID/Multicodecs/IPLD come into play altogether in this case.

I guess that the decision whether to classify these as IPLD codecs is up to you (aren't IPLD codecs using multicodecs anyway? why does it even matter if we specify them as one codec or another? I'm afraid I don't really understand the relationships between the different (many) moving parts in the multiformat/ipld specs). The discussions that were made on the topic were not very clear from our side, not to mention that a significant part of them was conducted off-band without any documentation or resolution. 

Please advise, and in any case it would be nice to hop on a call and maybe say hello and work this out more quickly as we want to merge the new ENS public resolver changes.

**Comment by justelad** [comment]
@Stebalien, I had a talk with @zelig and we have concluded that we'd like to add a new `multihash` code for `bmt` (as Swarm hashes are actually binary merkle tree hashes, not `keccak256` hashes per-se).

You could [see](https://github.com/ethersphere/go-ethereum/pull/389/files) how we've added local support for that at the time in Swarm (as of a few months ago we have dropped usage of `multiformats`, but this will be reintroduced with the changes to the resolver).

**Comment by Stebalien** [comment]
Sorry, github notifications tend to fall through the cracks. This looks good to me.

With respect to IPLD: IPLD is really just a (poorly documented...) attempt at unifying all merkle formats into a single system. It does so by using CIDs (content identifiers) which are `(multicodec, multihash)` tuples (where the multicodec == "ipld codec"). This allows us to point to an object in a merkle dag while recording the _type_ of merkledag we're pointing to.

TL;DR: IPLD is to merkledags as Multihash is to hash functions.
